### ISSUE TITLE
Difference in Qscores between BAM alignment and guppy summary.txt input

### ISSUE THREAD

**Issue by jfhaaijer** [wdecoster/NanoPlot#183]
Hello,

I've noticed a difference in the Q scores when I run NanoPlot with a BAM alignment input vs. a sequencing_summary.txt file from Guppy. I used the same samples as input, so i expected the quality scores to be more similar. Do know what could be causing this difference?

The samples are Deformed Wing Virus reads, basecalled with a model that was trained with [Taiyaki](https://github.com/nanoporetech/taiyaki). I also adjusted the Q scores for the model with Taiyaki's [calibrate_qscores_by_read.py](https://github.com/nanoporetech/taiyaki/blob/master/misc/calibrate_qscores_byread.py) script. 

### Sequencing_summary.txt
[NanoStats.txt](https://github.com/wdecoster/NanoPlot/files/4717782/NanoStats.txt)
![LengthvsQualityScatterPlot_kde](https://user-images.githubusercontent.com/55389441/83532817-11250800-a4ef-11ea-8952-2409bbd0d519.png)

### BAM Alignment
[NanoStats.txt](https://github.com/wdecoster/NanoPlot/files/4717798/NanoStats.txt)
![LengthvsQualityScatterPlot_kde](https://user-images.githubusercontent.com/55389441/83532879-2a2db900-a4ef-11ea-9d05-b6233e7f1e19.png)

**Comment by wdecoster** [comment]
There are a couple of differences that could contribute to a discrepancy between the summary and the bam, but not to this extent. For example, NanoPlot with a bam file will ignore both entirely unmapped reads and the parts of the reads which do not align (soft/hard clipped in the CIGAR). Also, the sequencing_summary file contains a Q-score as estimated from the basecaller, while based on the BAM file the average Q-score is computed from each Q-score of each base, which (at least in the past) wasn't exactly the same.

However, having a lot of reads with Q-score between 25 and 30 looks really suspicious - as that is very high for nanopore data. But I have never used Taiyaki, and neither needed calibration. Maybe @tmassingham-ont has some insights here? Could you show (from the NanoPlot --bam output) the plot comparing the percent identity with the Q-scores? That should at least give us an empirical idea of how accurate your reads really are.

**Comment by jfhaaijer** [comment]
Thanks for the quick reply! 
Here is the percent identity vs Q-scores plot:
![PercentIdentityvsAverageBaseQuality_kde](https://user-images.githubusercontent.com/55389441/83570694-caeb9b00-a526-11ea-96b6-24d3e4858475.png)

I've read on your blog that the percent identity plot should show the error rate matching the Q-score and wanted to use it to verify my basecalling model. But as you can see, they don't align as they should. However it looks like the Q-scores from the sequencing_summary.txt match the percent  identity from this plot a lot better. 

The high percent identity is probably a result of the basecalling model that was trained with Taiyaki. The model was trained on the samples and a reference genome was used as a ground truth. The resulting alignment, where the basecalled samples were aligned with the same reference genome, aligns quite well. 
![AlignedReadlengthvsSequencedReadLength_dot](https://user-images.githubusercontent.com/55389441/83572745-6599a900-a52a-11ea-8a3e-7c6443d996a9.png)

**Comment by tmassingham-ont** [comment]
Hello @jfhaaijer .  While I'd love to claim that that Q30 is exactly what you should expect, it looks like the calibration has not worked.  I've asked the author of the script to look into what has happened, since the adjusted scores should match the empirical values calculated by NanoPlot.

It would be useful to know how you basecalled the reads, please?

**Comment by jfhaaijer** [comment]
@tmassingham-ont, these are the steps i have done:
1. Basecall with guppy
`guppy_basecaller -i samples -s basecalled_reads -c rna_r9.4.1_70bps_hac.cfg --device cuda:0`

2. Map with Minimap
`minimap2 -ax splice -uf -k14 -t 32 reference.fasta basecalled_reads/*.fastq | samtools view -b -S -T reference.fasta- > mapped_basecalls.bam`

3. Prepare mapped reads
- `get_refs_from_sam.py reference.fasta mapped_basecalls.bam --min_coverage 0.8 > read_references.fasta --reverse`
- `generate_per_read_params.py --jobs 32 samples > read_params.tsv`
- `prepare_mapped_reads.py  --jobs 32 samples read_params.tsv mapped_reads.hdf5 taiyaki_walkthrough/pretrained/r941_rna_minion.checkpoint read_references.fasta`

4. Train a new model
- `train_flipflop.py --device cuda:0 taiyaki/models/mGru_flipflop.py mapped_reads.hdf5`
- `dump_json.py training/model_final.checkpoint > model.json`

5. Basecall with new model
`guppy_basecaller -i samples -s basecalled_reads_2 -c rna_r9.4.1_70bps_hac.cfg --device cuda:0`

6. Calibrate Q-scores
- `guppy_aligner -i basecalled_read_2 -s aligned_reads --align_ref reference.fasta | samtools sort --reference reference.fasta -o aligned_basecalls.bam`
- `taiyaki/misc/calibrate_qscores_byread.py --input_directory basecalled_reads_2`
![qscore_calibration](https://user-images.githubusercontent.com/55389441/83619864-bf32bf80-a58c-11ea-9248-338c04795ff8.png)


7. Basecall with adjusted Q-score scale and offset
`guppy_basecaller -i samples -s basecalled_reads_3 -c rna_r9.4.1_70bps_hac.cfg --qscore_offset -49.61 --qscore_scale 9.11 -m model.json --device cuda:0`

  `

**Comment by aspitaleri** [comment]
following

**Comment by tmassingham-ont** [comment]
Since the model was over-trained to a single sequence, we think the Q30 alignments are plausible and these agree with both the Taiyaki calibration plot and the percent ID calculated by NanoPlot.
The values for the 'Average read quality' and 'Average base quality' agree with the range of the uncalibrated quality values (x-axis of the calibration graph) -- as @wdecoster says these as calculated directly from the quality values in the BAM, I suspect the BAM file contains uncalibrated qualities.
